src/algorithms/models/vertex.ts

```diff
@@ -59,25 +57,29 @@ class Vertex {
   /**
    * Add a neighbor with edge
    */
-  addNeighbor(costs: EdgeCost, vertex: Vertex): void {
-    vertex.incrInDegree();
-    this.neighbors.add({
-      costs,
-      vertex
-    });
+  addNeighbor(edge: Edge): void {
+    // check if there exists the same edge already
+    for (const neighbor of this.neighbors) {
+      if (Edge.equal(neighbor, edge)) {
+        return;
+      }
+    }
+    edge.vertex.incrInDegree();
+    this.neighbors.add(edge);
   }
 
   /**
    * Get cost to neighbor. Returns undefined if neighbor not found.
-   * @param vertex : neighbor
+   * @param vertex: Vertex neighbor
    * @return EdgeCost
    */
   costTo(vertex: Vertex): EdgeCost {
     for (const neighbor of this.neighbors) {
-      if (neighbor.vertex === vertex) {
+      if (neighbor.vertex.id === vertex.id) {
```

Comment:
Why are you comparing IDs instead of the objects? Is there a possibility there are two objects representing one vertex at any given moment?